### PR TITLE
feat(core): Upgrade to React v16

### PR DESCRIPTION
**Pull request by timburgess**
## React Boilerplate

This PR extends from the Enzyme 3 upgrade PR.
- [x] Upgrades to React 16.0 and updates related dependencies
- [x] Upgrades Enzyme 3 adapter to 16 and adds shim required for React 16 testing
- [x] Adds ErrorBoundary component and uses in App

Build size goes from 690kB -> 630kB.

I'm not sure exactly if/why/how we should test ErrorBoundary.

Before opening a pull request, please ensure:

- [x] You have followed our [**contributing guidelines**](https://github.com/react-boilerplate/react-boilerplate/blob/master/.github/CONTRIBUTING.md)
- [x] double-check your branch is based on `dev` and targets `dev` 
- [x] Pull request has tests (we are going for 100% coverage!)
- [x] Code is well-commented, linted and follows project conventions
- [ ] Documentation is updated (if necessary)
- [x] Description explains the issue/use-case resolved and auto-closes related issues

Be kind to code reviewers, please try to keep pull requests as small and focused as possible :)

**IMPORTANT**: By submitting a patch, you agree to allow the project
owners to license your work under the terms of the [MIT License](https://github.com/react-boilerplate/react-boilerplate/blob/master/LICENSE.md).

### CODE REVIEW

**Comment by Droow** [comment]
I actually wrote a test for your error boundary and the tests are now at 100% coverage again. The test passes however it produces unnecessary output to console:
vyber_044

Is there any way to disable it or it is ok to show this error and console log among the test results?

**Comment by Droow** [comment]
Also I have added options to enable/disable showing and logging stack trace and support for i18n.
```javascript
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import messages from './messages';

export default class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false, error: null };
  }

  componentDidCatch(info) {
    // Display fallback UI
    this.setState({ hasError: true, error: info });
    if (this.props.logStackTrace) {
      console.warn(info); // eslint-disable-line no-console
    }
  }

  renderErrorStackTrace() {
    // Display stack trace only in dev environment
    if (process.env.NODE_ENV === 'development') {
      return (<pre>{this.state.error.stack}</pre>);
    }
    return null;
  }

  render() {
    if (this.state.hasError) {
      // You can render any custom fallback UI
      return (
        <div>
          <h1><FormattedMessage {...messages.title} /></h1>
          {this.props.showStackTrace && this.renderErrorStackTrace()}
        </div>
      );
    }
    return this.props.children;
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.node.isRequired,
  showStackTrace: PropTypes.bool,
  logStackTrace: PropTypes.bool,
};

ErrorBoundary.defaultProps = {
  showStackTrace: false,
  logStackTrace: false,
};
```
WIth obvious usage:
```javscript
<ErrorBoundary showStackTrace logStackTrace>
...
</ErrorBoundary>
```

**Comment by timburgess** [comment]
Thanks @Droow I've refactored `ErrorBoundary` to match your changes.

**Comment by justingreenberg** [comment]
@timburgess i'm thinking we should just use https://github.com/bvaughn/react-error-boundary by react core team member.. the api is nice and simple

since this is redux, it may also be useful to dispatch an error action... using a watcher saga, we can log the error to console in development or post enriched data to a remote service in production 

i haven't had the time to pull down code and test... thoughts?

@timburgess @Droow @Dattaya @KarandikarMihir

**Comment by timburgess** [comment]
@justingreenberg Do you mean import as a dependency or just use the code as the basis for `/components/ErrorBoundary`?

I'm imagining the latter as I envisage that ppl will want to customize ErrorBoundary to do different things. On my projects, I'm probably looking at hooking the ErrorBoundary catch into logrocket or trackjs.

**Comment by justingreenberg** [comment]
>On my projects, I'm probably looking at hooking the ErrorBoundary catch into logrocket or trackjs.

@timburgess absolutely! trackjs and logrocket are a little different... they hook onto window object and intercept console logging... with error boundaries we can handle react errors before they bubble up to window more gracefully inside application... logging to console is in itself a side effect, so should be isolated in either case and imo probably not a great practice in production environments

i'm not sure we would get much benefit from extracting components from `react-error-boundary` package, the `onError` hook pretty much leaves handling in userland.. i'm better at explaining by code so here's the general idea but very quickly written and not tested, just a brain dump....

simple action creator..
```js
// containers/App/actions.js
export function handleError({ error, componentStack }) {
  return {
    type: APP_ERROR,
    payload: { error, componentStack }
  };
}
```

wire into redux...
```js
// containers/App/index.js
// ...
import {connect} from 'redux';
import ErrorBoundary from 'react-error-boundary';
import {handleError} from 'containers/App/actions.js';

// connect to store, bind action creator
export default connect(null, { handleError })(({ handleError }) => {
  <ErrorBoundary
    onError={handleError}
    FallbackComponent={/* some fallback uh oh component */}>
    <Header />
    // ...
  </ErrorBoundary>
})
```

spawn a watcher saga to handle side effects... ie, logging to external service
```js
// containers/App/saga.js
import myTracker from "utils/tracker";

export default function* errorHandler() {
  yield takeLatest(APP_ERROR, function* ({ error, componentStack }) {
    const metadata = yield select(getEnrichmentDataSelector());
    myTracker.track({
      /* ... */
    });
  });
}
```

*now side effects are contained*
likewise, the reducer should set the application state (if you're not using fallback component api)

---

alternatively, since this boundary encompasses scope, we could use simple redux middleware to isolate effects, for example:

```js
// middleware
import myTracker from 'utils/tracker';

const errorMiddleware = store => next => action => {
  // handle errors from react error boundary
  if (action.type === APP_ERROR) {
    myTracker.track({ type: "REACT_ERROR" /* ... */ });
    return next(action);
  }

  // handle errors from invoking action creators
  try {
    return next(action);
  } catch (err) {
    myTracker.track({ type: "REDUX_ERROR" /* ... */ });
  }
};

```

---

again, just trying to brainstorm ideas... please feel free to provide feedback! i'm totally open to ideas, or please let me know if i'm making some massive oversight as i haven't had the time to test this or create demo implementation

thank you!

**Comment by justingreenberg** [comment]
> Also, FWIW react-error-boundary has a dependency on react ^16.0.0-alpha.13.

no worries on this peer... semver resolves pre-releases with caret to corresponding release :)

**Comment by timburgess** [comment]
I've pushed up some more changes which include a simple `ErrorFallback` with i18n and a simple saga for the error handling (not sure atm how to do something generic for err handling).

The part I'm struggling with is binding the `handleError` action creator. Do we not want people to be able to do something like:
```
        <ErrorBoundary
          onError={handleError}
          FallbackComponent=<ErrorFallback />
        >
          ...
        <ErrorBoundary />
```
wherever they wish in their component hierarchy and thus App uses some simple to understand binding code.

File: app/components/ErrorFallback/index.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Use this where required in React 16
+ */
+import React from 'react';
+import { FormattedMessage } from 'react-intl';
+
+import messages from './messages';
+
+function ErrorFallback() {
+  return <FormattedMessage {...messages.error} />;
```
Comment (blling) [inline app/components/ErrorFallback/index.js]: The message key should be a param ?

Comment (justingreenberg) [reply app/components/ErrorFallback/index.js]: right now there is a `defaultMessage` assigned here https://github.com/react-boilerplate/react-boilerplate/pull/1969/files#diff-329cc2749c47df1ee7d353ddde8039b8R11

**Comment by justingreenberg** [comment]
ok i pushed up some updates, all tests are passing just gotta get coverage above 98% threshold

also, i am dropping support for node v5 and v6 for now since our internal scripts to not support it anyway, eg https://travis-ci.org/react-boilerplate/react-boilerplate/jobs/292190822#L510

if anyone wants to make these scripts backward-compatible we can create an issue for that

File: app/components/Footer/index.js [hunk 2]
```diff
@@ -23,6 +23,13 @@ function Footer() {
           }}
         />
       </section>
+      <button
+        onClick={() => {
+          throw new Error('uh oh!');
```
[the remark this answers is not in the document]
Comment (timburgess) [reply app/components/Footer/index.js]: Hmmm. No doesn't work for me - just seems to flow through to a console error.

**Comment by justingreenberg** [comment]
@blling please see https://github.com/react-boilerplate/react-boilerplate/pull/1969/commits/73d0d2c4238990ee42484ac14d34ec8036dc5909 and https://github.com/react-boilerplate/react-boilerplate/pull/1969/commits/dacfdd094975ac1e95139c0d9f2c9f5ba08c10fb :)

FYI—right now there are a few minor blocking issues on merging this - 
1. test coverage
2. extract-intl script is not generating translations
3. confirm error boundary is working with a demo

any help or insight would be appreciated!

**Comment by timburgess** [comment]
@justingreenberg Thanks for the issues clarification. I'll find some time on the weekend to see if we can get this done.

File: app/components/Footer/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaearon) [reply app/components/Footer/index.js]: Error boundaries aren't supposed to catch errors in event handlers. Only in render and lifecycles.

[the remark this answers is not in the document]
Comment (justingreenberg) [reply app/components/Footer/index.js]: @gaearon thank you for clarification

File: app/containers/App/index.js [hunk 3]
```diff
@@ -42,15 +45,9 @@ export const App = ({ handleError: errorHandler }) => (
...
-        name="description"
-        content="A React.js Boilerplate application"
-      />
+      <meta name="description" content="A React.js Boilerplate application" />
     </Helmet>
-    <ErrorBoundary
-      onError={errorHandler}
-      FallbackComponent={<ErrorFallback />}
-    >
+    <ErrorBoundary onError={errorHandler} FallbackComponent={<ErrorFallback />}>
```
Comment (blling) [inline app/containers/App/index.js]: `ErrorBoundary` support HOC in version `1.1.0`, Use a HOC is better.

**Comment by blling** [comment]
I have send a pr [here](https://github.com/timburgess/react-boilerplate/pull/1) , for:
1. test coverage
3. confirm error boundary should have working demo

cc @timburgess @justingreenberg

File: app/containers/App/index.js [hunk 4]
```diff
@@ -7,15 +7,28 @@
...
 
+// Async bundles (code split-points)
 import HomePage from 'containers/HomePage/Loadable';
 import FeaturePage from 'containers/FeaturePage/Loadable';
 import NotFoundPage from 'containers/NotFoundPage/Loadable';
+
+// Dumb components
 import Header from 'components/Header';
 import Footer from 'components/Footer';
+import ErrorFallback from 'components/ErrorFallback';
```
Comment (SimonArdrey) [inline app/containers/App/index.js]: `react-error-boundary` contains `ErrorBoundaryFallbackComponent` should we consider reusing that?

File: app/containers/App/index.js [hunk 5]
```diff
@@ -26,22 +39,37 @@ const AppWrapper = styled.div`
...
-      </Helmet>
+export const App = ({ handleError: errorHandler }) => (
+  <AppWrapper>
+    <Helmet
+      titleTemplate="%s - React.js Boilerplate"
+      defaultTitle="React.js Boilerplate"
+    >
+      <meta name="description" content="A React.js Boilerplate application" />
+    </Helmet>
+    <ErrorBoundary onError={errorHandler} FallbackComponent={<ErrorFallback />}>
```
Comment (SimonArdrey) [inline app/containers/App/index.js]: Should we have `ErrorBoundary` here or above the `AppWrapper`? I think any errors in components higher in the tree will not be caught by this boundary.

File: app/containers/App/reducer.js [hunk 6]
```diff
@@ -41,9 +44,7 @@ function appReducer(state = initialState, action) {
         .set('loading', false)
         .set('currentUser', action.username);
     case LOAD_REPOS_ERROR:
-      return state
-        .set('error', action.error)
-        .set('loading', false);
+      return state.set('error', action.error).set('loading', false);
```
Comment (SimonArdrey) [inline app/containers/App/reducer.js]: I think that state props should not use multiple datatypes. In this reducer `error` is sometimes set to `true` and sometimes set to the `action.error`

Should we store the result of the `ErrorBoundary` in the redux state at all? Is it within the scope of pushing through a React 16 upgrade?

File: app/containers/App/index.js (hunk 4 above)
Comment (timburgess) [reply app/containers/App/index.js]: Thanks for pointing that out. I'll take a look.

File: app/containers/App/index.js (hunk 5 above)
Comment (timburgess) [reply app/containers/App/index.js]: Yeah, probably higher up.

**Comment by mxstbr** [comment]
Deploy preview failed.

Built with commit 4ea5c1949a13d7bab7e48290162a412829059051

https://app.netlify.com/sites/react-boilerplate/deploys/59fd0aba4c4b9340d0cb0eee

**Comment by timburgess** [comment]
I eventually want `i18n` BoundaryError customization too but at present I want React 16 more. And

screenshot from 2017-11-04 13-43-13

is pretty universal.

If you want to have a custom ErrorFallback component in your projects, it's pretty simple to create one and reference that instead.

**Comment by timburgess** [comment]
Travis CS is failing just on Node 6. Not exactly sure why.

**Comment by mradziwon** [comment]
Is node v6 the only problem left before merging this pr? I would love to use react 16 for new project.

**Comment by blling** [comment]
I think it is not ok and still need some improvments and review.

**Comment by bytewiz** [comment]
Need this upgrade, anything I can do to help get it out there faster?

**Comment by gretzky** [comment]
@bytewiz and all- i've successfully updated to react v16 on my own spin off of react-boilerplate ([yeti](https://github.com/gretzky/yeti)). later on today i can start a new PR with this update and see what it'll take to push it forward.

**Comment by blling** [comment]
@gretzky Great. 
I think this `story` should be split into `sub-tasks`, here is some `sub-tasks` i could imagine
1. Update dependecies and resolve related issues.
2. Add `Logging intercept`, error log should be tracked. Such as @justingreenberg suggested. 
3. Add `ErrorBoundary`(HOC) and add a working demo
4. Add `babel-plugin-transform-react-jsx-source`, see 73d0d2c and dacfdd0

Hope i could do something on this `story`.

**Comment by blling** [comment]
@timburgess Should we close this PR? 
The ErrorBoundry and Error Logging Intercept should do in another PR. 
Maybe they need more discussion for the final design.

**Comment by timburgess** [comment]
Fine with me. My weekends have dissolved time-wise in Dec unfortunately.

**Comment by blling** [comment]
OK. I will close this for now, thanks!

**Pull request closed without merging**